src/modules/job-manager/plugins/alloc-bypass.c

```diff
@@ -21,33 +21,116 @@
 #include <flux/core.h>
 #include <flux/jobtap.h>
 
+static void alloc_continuation (flux_future_t *f, void *arg)
+{
+    flux_plugin_t *p = arg;
+    flux_jobid_t *idptr = flux_future_aux_get (f, "jobid");
+
+    if (flux_future_get (f, NULL) < 0) {
+        flux_jobtap_raise_exception (p,
+                                     *idptr,
+                                     "alloc", 0,
+                                     "failed to commit R to kvs: %s",
+                                      strerror (errno));
+        goto done;
+    }
+    if (flux_jobtap_event_post_pack (p,
+                                     *idptr,
+                                     "alloc",
+                                     "{s:b}",
+                                     "bypass", true) < 0)
+        flux_jobtap_raise_exception (p,
+                                     *idptr,
+                                     "alloc", 0,
+                                     "failed to post alloc event: %s",
+                                     strerror (errno));
+done:
+    flux_future_destroy (f);
+}
+
+static int alloc_start (flux_plugin_t *p,
+                        flux_jobid_t id,
+                        const char *R)
+{
+    flux_t *h;
+    char key[64];
+    flux_future_t *f = NULL;
+    flux_kvs_txn_t *txn = NULL;
+    flux_jobid_t *idptr = calloc (1, sizeof (*idptr));
+
+    if (!idptr || !(h = flux_jobtap_get_flux (p))) {
+        free (idptr);
+        return -1;
+    }
+
+    *idptr = id;
+
+    if (flux_job_kvs_key (key, sizeof (key), id, "R") < 0
+        || !(txn = flux_kvs_txn_create ())
+        || flux_kvs_txn_put (txn, 0, key, R) < 0
+        || !(f = flux_kvs_commit (h, NULL, 0, txn))
+        || flux_future_aux_set (f, "jobid", idptr, free) < 0
+        || flux_future_then (f, -1, alloc_continuation, p)) {
+        flux_kvs_txn_destroy (txn);
+        flux_future_destroy (f);
+        return -1;
+    }
```

Comment:
idptr is leaked if any of the chained checks fail up to and including the aux_set.

Reply:
Good catch. Passing a jobid to a callback (future or message) may be a common case. I wonder if we could make this easier somehow.. kind of annoying to have to allocate memory for this.

Reply:
Maybe it would be useful to have a handle for a job that is opaque to the plugin, but actually references a struct job?

Reply:
Uh yeah, I was wishing we'd gone that way before.. it would be a change to the general approach of the API, but perhaps a lot more user friendly given the evolution of the api..